Build fixes, especially around cmake+monolithic

jbfove opened the pull request:
Was integrating wxWidgets into another larger cmake project, using the monolithic build, and ran into a number of build issues. Though not all of them are necessarily related to the monolithic build

Contributing back the fixes back so the next person doesn't need to solve them all over again.

No tests are added since these are build fixes, no new or changed features.

This is my first MR for wx so I'm more than happy to rework things or make any changes as needed.

jbfove committed:
Fix extraneous parens warning

jbfove committed:
Fix use of wxDateTime without check for wxUSE_DATETIME

jbfove committed:
Fix warning with missing override

jbfove committed:
Fix uses of wxPalette without check for wxUSE_PALETTE

jbfove committed:
Fix use of RECT in wxClientDisplayRect without including windows headers

jbfove committed:
Fix monolithic build in cmake (wxBUILD_MONOLITHIC)

- Functions that set wxMONO_* vars need to set them in parent scope, from macros instead of functions (functions would need grandparent scope)
- Fix a conflict with xml library linking Mono which made core include the wrong png headers
- Fix setting of NOPCH on windows to go through the wx_lib_compile_definitions instead of target_compile_definitions so it carries to monolithic build
- Allow use of precompiled header in MSVC via calling wx_finalize_lib(mono)
- Introduce wxMONO_NONCOMPILED_CPP_FILES to deal with set_source_files_properties only applying within the current CMakeLists file
- Fix demos building against monolithic library by using wx_exe_link_libraries instead of target_link_libraries (similar to what the tests already do)

Closes #18074

MaartenBent commented on build/cmake/files.cmake:
```diff
@@ -588,7 +588,6 @@ set(BASE_CMN_HDR
     wx/textbuf.h
     wx/textfile.h
     wx/thread.h
-    wx/thrimpl.cpp
```
`files.cmake` is auto-generated, so the build error caused by this file should be solved in another way.
And preferable in it's own commit.

jbfove replied:
Thanks, I missed the autogenerated note at the top. I'll take a stab at fixing this another way, or remove it from the MR if I can't easily find a fix.

jbfove replied:
Found the root issue here: `set_source_files_properties(... HEADER_FILE_ONLY TRUE)` is used to make the .cpp "headers" not compile. However the `mono` target is created in a different scope so it couldn't read the properties set. I introduced a new variable `wxMONO_NONCOMPILED_CPP_FILES` to pass these upscope so the `mono` target can set this within it's own scope. Fix is pushed.

MaartenBent reviewed (commented):
Looks great.

Monolithic build now works fine in VS 2017. But it takes ages to build because pch are not enabled. This could be fixed in `build\cmake\lib\CMakeLists.txt` with:

```diff
@@ -87,6 +92,7 @@ if(wxBUILD_MONOLITHIC)
             target_compile_definitions(mono ${vis} ${wxMONO_DEFINITIONS_${vis}})
         endif()
     endforeach()
+    wx_finalize_lib(mono)
 endif()
 
 # Propagate variable(s) to parent scope
```

There is also a ticket for this problem: https://trac.wxwidgets.org/ticket/18074

jbfove commented:
Force pushed 3 changes:
- Proper fix for the files.cmake issue (per comment above)
- Fixed samples building against the mono target(by using `wx_exe_link_libraries` instead of `target_link_libraries`)
- Added the trac number to the commit message (note: all my monolithic-related fixes are in one commit since it doesn't work without all of them, but I can split them into individual commits, where only the last one will work, if that's preferred)
- Added `wx_finalize_lib(mono)` per your suggestion, speed up the windows builds (but currently on OSX so I haven't tested this yet)

vadz commented:
Thanks for the fixes! I can't meaningfully review the CMake part, but Maarten has already review that, and the rest looks good to me, so merged now.

Thanks again!

LegalizeAdulthood commented:
CMake part looks good to me.
